internal/gitutil/gitutil.go

```diff
@@ -378,19 +394,24 @@ loop:
 		// commit sha.
 		validFullSha := s == strings.TrimSpace(rr.Stdout)
 		_, resolved := gur.ResolveRef(s)
-
+		// fetchedRefWithURI is the cache key created from repo URI SHA and ref
+		fetchedRefWithURI := fmt.Sprintf("%s-%s", uriSha, s)
+		_, fetched := gur.fetchedRefs[fetchedRefWithURI]
 		switch {
 		case resolved || validFullSha:
 			// If the ref references a branch or a tag, or is a valid commit
-			// sha, we can fetch just a single commit.
-			if _, err := gitRunner.RunVerbose(ctx, "fetch", "origin", "--depth=1", s); err != nil {
-				AmendGitExecError(err, func(e *GitExecError) {
-					e.Repo = uri
-					e.Command = "origin"
-					e.Ref = s
-				})
-				return "", errors.E(op, errors.Git, fmt.Errorf(
-					"error running `git fetch` for ref %q: %w", s, err))
+			// sha and has not already been fetched, we can fetch just a single commit.
+			if !fetched {
+				if _, err := gitRunner.RunVerbose(ctx, "fetch", "origin", "--depth=1", s); err != nil {
+					AmendGitExecError(err, func(e *GitExecError) {
+						e.Repo = uri
+						e.Command = "fetch"
+						e.Ref = s
+					})
+					return "", errors.E(op, errors.Git, fmt.Errorf(
+						"error running `git fetch` for ref %q: %w", s, err))
+				}
+				gur.fetchedRefs[fetchedRefWithURI] = true
```

Comment:
Do we need to update it in case of `default` case of this switch statement (below) ?

Reply:
Yeah we can expand caching to default case since we don't need to rerun after initial fetch.